intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,525 @@
+    - Broadcasting operations
+    - Flattening and unflattening dimensions
+
+Finally, we'll put this into practice by writing a multi-head attention module
+using named tensors.
+
+Named tensors in PyTorch are inspired by and done in collaboration with
+`Sasha Rush <https://tech.cornell.edu/people/alexander-rush/>`_.
+The original idea and proof of concept were proposed in his
+`January 2019 blog post <http://nlp.seas.harvard.edu/NamedTensor>`_.
+
+Basics: named dimensions
+========================
+
+PyTorch now allows Tensors to have named dimensions; factory functions
+now take a new `names` argument that associates a name with each dimension.
+This works with most factory functions, such as
+
+- `tensor`
+- `empty`
+- `ones`
+- `zeros`
+- `randn`
+- `rand`
+
+Here we construct a tensor with names:
+"""
+
+import torch
+imgs = torch.randn(1, 2, 2, 3 , names=('N', 'C', 'H', 'W'))
+print(imgs.names)
+
+######################################################################
+# Unlike in
+# `the original named tensors blogpost <http://nlp.seas.harvard.edu/NamedTensor>`_,
+# named dimensions are ordered: ``tensor.names[i]`` is the name of the ``i`` th dimension of ``tensor``.
+#
+# There are two ways rename a ``Tensor``'s dimensions:
+
+# Method #1: set .names attribute
+imgs.names = ['batch', 'channel', 'width', 'height']
+print(imgs.names)
+
+# Method #2: specify new names (note: this changes names out-of-place)
+imgs = imgs.rename(channel='C', width='W', height='H')
+print(imgs.names)
+
+######################################################################
+# The preferred way to remove names is to call ``tensor.rename(None)``:
+
+imgs = imgs.rename(None)
+print(imgs.names)
+
+######################################################################
+# Unnamed tensors (tensors with no named dimensions) still work as normal and do
+# not have names in their ``repr``.
+
+unnamed = torch.randn(2, 1, 3)
+print(unnamed)
+print(unnamed.names)
+
+######################################################################
+# Named tensors do not require that all dimensions be named.
+
+imgs = torch.randn(3, 1, 1, 2, names=('N', None, None, None))
+print(imgs.names)
+
+######################################################################
+# Because named tensors coexist with unnamed tensors, we need a nice way to
+# write named tensor-aware code that works with both named and unnamed tensors.
+# Use ``tensor.refine_names(*names)`` to refine dimensions and lift unnamed dims
+# to named dims. Refining a dimension is defined as a "rename" with the following
+# constraints:
+#
+# - A ``None`` dim can be refined to have any name
+# - A named dim can only be refined to have the same name.
+
+imgs = torch.randn(3, 1, 1, 2)
+print(imgs.refine_names('N', 'C', 'H', 'W'))
+
+# Coerces the last two dims to 'H' and 'W'. In Python 2, use the string '...' instead of ...
+print(imgs.refine_names(..., 'H', 'W').names)
+
+def catch_error(fn):
+    try:
+        fn()

```

Comment:
Let's change this to 

```
fn()
assert False
```
Because whenever we run `catch_error`, there should be an error